File: core/rpc_manager.py
class RPCManager:
    def __init__(self, rpc_entries):
        self.rpcs = rpc_entries
        self.index = 0
        self.fail_count = {}

        for i in range(len(rpc_entries)):
            self.fail_count[i] = 0

    def get(self):
        if not self.rpcs:
            return None

        entry = self.rpcs[self.index % len(self.rpcs)]
        self.index += 1
        return entry

    def mark_fail(self, entry):
        try:
            idx = self.rpcs.index(entry)
            self.fail_count[idx] += 1
        except:
            pass

    def mark_success(self, entry):
        try:
            idx = self.rpcs.index(entry)
            self.fail_count[idx] = 0
        except:
            pass

File: core/rpcManager.js
class RPCManager {
  constructor(rpcEntries) {
    this.rpcs = rpcEntries;
    this.index = 0;
    this.failCount = new Map();

    for (let i = 0; i < rpcEntries.length; i += 1) {
      this.failCount.set(i, 0);
    }
  }

  get() {
    if (this.rpcs.length === 0) {
      return null;
    }

    const entry = this.rpcs[this.index % this.rpcs.length];
    this.index += 1;
    return entry;
  }

  markFail(entry) {
    const index = this.rpcs.indexOf(entry);
    if (index >= 0) {
      this.failCount.set(index, (this.failCount.get(index) || 0) + 1);
    }
  }

  markSuccess(entry) {
    const index = this.rpcs.indexOf(entry);
    if (index >= 0) {
      this.failCount.set(index, 0);
    }
  }
}

module.exports = { RPCManager };